compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -597,22 +548,67 @@ impl<'tcx> Validator<'_, 'tcx> {
         }
     }
 
+    fn validate_ref(&self, kind: BorrowKind, place: PlaceRef<'tcx>) -> Result<(), Unpromotable> {
+        if let BorrowKind::Mut { .. } = kind {
+            let ty = Place::ty_from(place.local, place.projection, self.body, self.tcx).ty;
+
+            // In theory, any zero-sized value could be borrowed
+            // mutably without consequences. However, only &mut []
+            // is allowed right now, and only in functions.
+            if let ty::Array(_, len) = ty.kind() {
+                match len.try_eval_usize(self.tcx, self.param_env) {
+                    Some(0) => {}
+                    _ => return Err(Unpromotable),
+                }
+            } else {
+                return Err(Unpromotable);
+            }
+        }
+
+        // HACK(eddyb) this should compute the same thing as
+        // `<HasMutInterior as Qualif>::in_projection` from
+        // `check_consts::qualifs` but without recursion.
+        let mut has_mut_interior = self.qualif_local::<qualifs::HasMutInterior>(place.local);
+        if has_mut_interior {
+            let mut place_projection = place.projection;
+            // FIXME(eddyb) use a forward loop instead of a reverse one.
+            while let &[ref proj_base @ .., elem] = place_projection {
+                // FIXME(eddyb) this is probably excessive, with
+                // the exception of `union` member accesses.
+                let ty = Place::ty_from(place.local, proj_base, self.body, self.tcx)
+                    .projection_ty(self.tcx, elem)
+                    .ty;
+                if ty.is_freeze(self.tcx.at(DUMMY_SP), self.param_env) {
+                    has_mut_interior = false;
+                    break;
+                }
+
+                place_projection = proj_base;
+            }
+        }
+        if has_mut_interior {
+            return Err(Unpromotable);
+        }
+
+        Ok(())
+    }
+
     fn validate_rvalue(&self, rvalue: &Rvalue<'tcx>) -> Result<(), Unpromotable> {
         match *rvalue {
-            Rvalue::Cast(CastKind::Misc, ref operand, cast_ty) if self.const_kind.is_none() => {
+            Rvalue::Cast(CastKind::Misc, ref operand, cast_ty) if self.maybe_runtime() => {
```

Comment:
Pointer to int casts are not allowed in any const context. We should never promote these.

Reply:
I can remove it, but in the review comment you said you preferred not changing behavior without a crater run so I am not sure now what you'd prefer me to do.

Reply:
To me, this change is trivial, while the `validate_ref` refactor is not.